wrapper.cc

```diff
@@ -15,119 +15,133 @@
+            Nan::HandleScope scope;
+
+            v8::Local<v8::Value> argv[] = {
+                New<v8::Integer>(*reinterpret_cast<int*>(const_cast<char*>(data)))
+            };
+            progress->Call(1, argv);
+        }
+
+        // We have the results, and we're back in the event loop.
+        void HandleOKCallback () {
+            Isolate* isolate = Isolate::GetCurrent();
+            Nan::HandleScope scope;
+
+            Local<Array> results = New<Array>(contacts.size());
+            int i = 0;
+            for_each(contacts.begin(), contacts.end(), [&](Person* person) {
+                    Local<Object> contact = Object::New(isolate);
+                    fillPersonObject(isolate, contact, person);
+                    Nan::Set(results, i,  contact);
+                    i++;
+            });
+
+            Local<Value> argv[] = {results};
+            callback->Call(1, argv);
+        }
+    private:
+      Callback *progress;
+      vector<Person*> contacts;
+};
+
+// Asynchronous access to the `getContacts()` function
+NAN_METHOD(GetContacts) {
+    Callback *progress = new Callback(info[0].As<Function>());
+    Callback *callback = new Callback(info[1].As<Function>());
+
+    AsyncQueueWorker(new AddressBookWorker(callback, progress));
 }
 
-void GetMe(const FunctionCallbackInfo<Value>& args)
-{
-	AddressBook ab;
-	Isolate* isolate = args.GetIsolate();
+NAN_METHOD(GetMe) {
+    AddressBook ab;
+    Isolate* isolate = Isolate::GetCurrent();
+
+    Local<Object> me = Object::New(isolate);
+    fillPersonObject(isolate, me, ab.getMe());
 
-	Local<Object> me = Object::New(isolate);
-	fillPersonObject(isolate, me, ab.getMe());
-	
-	args.GetReturnValue().Set(me);
+    info.GetReturnValue().Set(me);
 }
 
-void ContactCount(const FunctionCallbackInfo<Value>& args)
-{
-	AddressBook ab;
-	Isolate* isolate = args.GetIsolate();
 
-	Local<Integer> c = Integer::New(isolate, ab.contactCount());
-	
-	args.GetReturnValue().Set(c);
-}
+NAN_METHOD(GetContact) {
+    int index = info[0]->Uint32Value();
 
-void GetContact(const FunctionCallbackInfo<Value>& args)
-{
-	AddressBook ab;
-	Isolate* isolate = args.GetIsolate();
-
-	if (args.Length() < 1) {
-		// Throw an Error that is passed back to JavaScript
-		isolate->ThrowException(Exception::TypeError(
-		String::NewFromUtf8(isolate, "Wrong number of arguments")));
-		return;
-	}
-
-	// Check the argument types
-	if (!args[0]->IsNumber()) {
-		isolate->ThrowException(Exception::TypeError(
-		String::NewFromUtf8(isolate, "Wrong argument type")));
-		return;
-	}
-
-	// Perform the operation
-	int64_t i = args[0]->IntegerValue();
-
-	Local<Object> contact = Object::New(isolate);
-	Person *p = ab.getContact(i);
-	fillPersonObject(isolate, contact, p);
-	delete p;
-
-	args.GetReturnValue().Set(contact);
+    AddressBook ab;
+    Isolate* isolate = Isolate::GetCurrent();
+
+    Local<Object> contact = Object::New(isolate);
+    fillPersonObject(isolate, contact, ab.getContact(index));
```

Comment:
This code crashes if an invalid index is given, which happens in example.js if access to contacts isn't granted. Should either check `index => 0 && index < ab.contactCount()` and/or that `ab.getContact()` doesn't return NULL.